server/src/main/java/com/linecorp/centraldogma/server/internal/api/GitHttpService.java

```diff
@@ -0,0 +1,187 @@
+import com.linecorp.armeria.common.stream.ByteStreamMessage;
+import com.linecorp.armeria.common.stream.StreamMessage;
+import com.linecorp.armeria.server.annotation.Get;
+import com.linecorp.armeria.server.annotation.Header;
+import com.linecorp.armeria.server.annotation.Param;
+import com.linecorp.armeria.server.annotation.Post;
+import com.linecorp.centraldogma.server.internal.api.auth.RequiresReadPermission;
+import com.linecorp.centraldogma.server.internal.storage.project.ProjectApiManager;
+
+/**
+ * A service that provides Git HTTP protocol.
+ */
+@RequiresReadPermission
+public final class GitHttpService {
+
+    private static final Logger logger = LoggerFactory.getLogger(GitHttpService.class);
+
+    private static final char[] HEX_DIGITS = "0123456789abcdef".toCharArray();
+
+    // TODO(minwoox): Add the headers in this class to Armeria.
+    private static final AggregatedHttpResponse CAPABILITY_ADVERTISEMENT_RESPONSE = AggregatedHttpResponse.of(
+            ResponseHeaders.builder(200)
+                           .add(HttpHeaderNames.CONTENT_TYPE, "application/x-git-upload-pack-advertisement")
+                           .add(HttpHeaderNames.CACHE_CONTROL, ServerCacheControl.REVALIDATED.asHeaderValue())
+                           .build(),
+            HttpData.ofUtf8(capabilityAdvertisement()));
+
+    // https://git-scm.com/docs/protocol-capabilities/
+    private static String capabilityAdvertisement() {
+        final StringBuilder sb = new StringBuilder();
+        sb.append("001e# service=git-upload-pack\n0000");
+        pktLine(sb, "version 2");
+        pktLine(sb, COMMAND_LS_REFS);
+        // Support limited options for now due to the unique characteristics of Git repositories in
+        // Central Dogma, such as having only a master branch and no tags, among other specifics.
+        pktLine(sb, COMMAND_FETCH + '=' + OPTION_WAIT_FOR_DONE + ' ' + OPTION_SHALLOW);
+        // TODO(minwoox): Migrate hash function https://git-scm.com/docs/hash-function-transition
+        pktLine(sb, "object-format=sha1");
+        sb.append("0000"); // flush-pkt https://git-scm.com/docs/protocol-v2/2.31.0#_packet_line_framing
+        return sb.toString();
+    }
+
+    // https://git-scm.com/docs/protocol-common#_pkt_line_format
+    static void pktLine(StringBuilder sb, String line) {
+        lineLength(sb, line.length() + 5);
+        sb.append(line).append('\n');
+    }
+
+    private static void lineLength(StringBuilder sb, int length) {
+        for (int i = 3; i >= 0; i--) {
+            sb.append(HEX_DIGITS[(length >>> (4 * i)) & 0xf]);
+        }
+    }
+
+    private final ProjectApiManager projectApiManager;
+
+    public GitHttpService(ProjectApiManager projectApiManager) {
+        this.projectApiManager = requireNonNull(projectApiManager, "projectApiManager");
+    }
+
+    // https://www.git-scm.com/docs/gitprotocol-http#_smart_clients
+    @Get("/{projectName}/{repoName}/info/refs")
```

Comment:
What do you think about using the path similar to what a user sees in their browser location field? i.e. `/projects/{projectName}/repos/{repoName}` ?

Reply:
If so, users need to do `git clone https://dogma.com/projects/foo/repos/repo.git` instead of `git clone https://dogma.com/foo/repo.git` because we cannot control the request URI of Git command. Do you still prefer this approach?

Reply:
I'm not very opinionated on this, so let's go with `/<proj>/<repo>.git` and change if we really have to.